webapp/tests/test_render.py

```diff
@@ -323,6 +326,88 @@ def test_render_input_target(self):
         self.assertEqual(data['target'], '12')
         self.assertEqual(data['input_target'], 'constantLine(12)')
 
+    def fetchData(a, b):
+        return [TimeSeries('abc.pqr', 1, 4, 1, [1, 2, 3, 4])]
+
+    @mock.patch('graphite.render.evaluator.fetchData', fetchData)
+    def test_render_input_target_for_invalid_user_input(self):
+        graphite_target = 'movingAverage(abc.pqr, 15min)'
+        url = reverse('graphite.render.views.renderView')
+        response = self.client.get(url, {
+                 'target': graphite_target,
+                 'format': 'json',
+                 'from': '07:01_20140226',
+                 'until': '08:01_20140226',
+        })
+        assert response.status_code == 400
+        data = json.loads(response.content)
+        self.assertEqual(
+            data['message'],
+            "Invalid target: {}. Error: int() argument must be a string or a number, not 'list'.".format(
+                graphite_target
+            )
+        )
+
+    @mock.patch('graphite.render.evaluator.fetchData', fetchData)
+    def test_render_input_target_for_non_existing_function(self):
+        graphite_target = 'nonExistingFunction(abc.pqr, 15min)'
+        url = reverse('graphite.render.views.renderView')
+        response = self.client.get(url, {
+                 'target': graphite_target,
+                 'format': 'json',
+                 'from': '07:01_20140226',
+                 'until': '08:01_20140226',
+        })
+        assert response.status_code == 400
+        data = json.loads(response.content)
+        self.assertEqual(
+            data['message'],
+            'Invalid target: {}. Error: "Unknown graphite function u\'nonExistingFunction\'".'.format(
+                graphite_target
+            )
+        )
+
+    @mock.patch('graphite.render.evaluator.fetchData', fetchData)
+    def test_render_input_target_for_invalid_from_format(self):
+        graphite_target = 'nonExistingFunction(abc.pqr, 15min)'
```

Comment:
might want to change this to a valid function.

Reply:
Rename the function names to valid ones.